src/transaction/transaction_manager.cpp

```diff
@@ -152,28 +183,37 @@ void TransactionManager::checkpointNoLock(main::ClientContext& clientContext) {
-    clientContext.getCatalog()->checkpoint(clientContext.getDatabasePath(),
-        clientContext.getVFSUnsafe());
-    // Log the checkpoint to the WAL and flush WAL. This indicates that all shadow pages and files(
-    // snapshots of catalog and metadata) have been written to disk. The part is not done is replace
-    // them with the original pages or catalog and metadata files.
-    // If the system crashes before this point, the WAL can still be used to recover the system to a
-    // state where the checkpoint can be redo.
-    wal.logAndFlushCheckpoint();
-    // Replace the original pages and catalog and metadata files with the updated/newly-created
-    // ones.
-    StorageUtils::overwriteWALVersionFiles(clientContext.getDatabasePath(),
-        clientContext.getVFSUnsafe());
-    clientContext.getStorageManager()->getShadowFile().replayShadowPageRecords(clientContext);
-    // Clear the wal, and also shadowing files.
-    wal.clearWAL();
-    clientContext.getStorageManager()->getShadowFile().clearAll(clientContext);
-    StorageUtils::removeWALVersionFiles(clientContext.getDatabasePath(),
-        clientContext.getVFSUnsafe());
+    try {
+        // Checkpoint node/relTables, which writes the updated/newly-inserted pages and metadata to
+        // disk.
+        clientContext.getStorageManager()->checkpoint(clientContext);
+        // Checkpoint catalog, which serializes a snapshot of the catalog to disk.
+        clientContext.getCatalog()->checkpoint(clientContext.getDatabasePath(),
+            clientContext.getVFSUnsafe());
+        // Log the checkpoint to the WAL and flush WAL. This indicates that all shadow pages and
+        // files( snapshots of catalog and metadata) have been written to disk. The part is not done
+        // is replace them with the original pages or catalog and metadata files. If the system
+        // crashes before this point, the WAL can still be used to recover the system to a state
+        // where the checkpoint can be redo.
+        wal.logAndFlushCheckpoint();
+        // Replace the original pages and catalog and metadata files with the updated/newly-created
+        // ones.
+        StorageUtils::overwriteWALVersionFiles(clientContext.getDatabasePath(),
+            clientContext.getVFSUnsafe());
+        clientContext.getStorageManager()->getShadowFile().replayShadowPageRecords(clientContext);
+        // Clear the wal, and also shadowing files.
+        wal.clearWAL();
+        clientContext.getStorageManager()->getShadowFile().clearAll(clientContext);
+        StorageUtils::removeWALVersionFiles(clientContext.getDatabasePath(),
+            clientContext.getVFSUnsafe());
+    } catch (std::exception& e) {
+        // since we want to rollback the checkpoint atomically
+        // we maintain ownership of the locks (which we will release when we are done handling the
+        // exception)
+        common::CheckpointException checkpointException(e);
```

Comment:
Can we not handle rollback here? instead of back to ClientContext? so lots of things can be simplified.

Reply:
that's probably a good idea

Reply:
updated